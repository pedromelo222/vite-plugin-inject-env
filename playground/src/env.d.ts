
interface SomeInterface {
  someText: string

}

interface ImportMetaEnv {
  readonly VITE_APP_BASE_API_URL: string
  readonly VITE_APP_BASE_API_SOCKET_URL: string
  readonly VITE_APP_BASE_AUTH_API: string
  readonly VITE_APP_JSON_ENV_BASE_API_URL: string
  readonly VITE_APP_JSON_ENV_BASE_API_SOCKET_URL: string
  readonly VITE_APP_JSON_ENV_BASE_AUTH_API: string
  readonly VITE_APP_MD_ENV_BASE_API_URL: string
  readonly VITE_APP_MD_ENV_BASE_API_SOCKET_URL: string
  readonly VITE_APP_MD_ENV_BASE_AUTH_API: string
  readonly VITE_APP_DOT_ENV_BASE_API_URL: string
  readonly VITE_APP_DOT_ENV_BASE_API_SOCKET_URL: string
  readonly VITE_APP_DOT_ENV_BASE_AUTH_API: string
}