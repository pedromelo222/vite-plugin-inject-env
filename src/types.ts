
export type RecordEnv = {
	[key in string]: any
}

export const importMeta = 'import.meta.env.VITE_APP_' as const


type ParsedKeyType<T extends string | symbol | number> = `${typeof importMeta}${Uppercase<T & string>}`

export type ParsedEnv<T extends RecordEnv> = {
	[Key in ParsedKeyType<keyof T>]: T[Key] | string
}

export interface InjectPluginConfig<T extends RecordEnv> {
	/**
	 * 
	 * javascript object Record<any, any>
	 * exemple: module: [env, otherEnv] - module: env   * 
	 * 
	 */
	module?: T | T[]
	/**z
	 * 
	 * Path of env file
	 * Accepted files: JSON, YML and MD 
	 * example: path: ['./src/env.json', './.otherEnv'] - path: './src/env.json'
	 * 
	 */
	path?: string | string[]
	/**
	 * 
	 * Flag to set if plugin must handle the creation of type definitions for the injected envs
	 * path taht typedefinitions are created: 'src/env.d.ts'
	 * default is true
	 * 
	 */
	typeDefinition?: boolean,
	/**
	 *  
	 * TypeDefinition path
	 * path location to store the enviroment type definition
	 *  
	 */
	//typeDefinitionPath?: string
}


