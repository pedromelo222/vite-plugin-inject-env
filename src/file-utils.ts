import { readFile } from 'node:fs/promises'

export function transformStringToObject(text: string) {
  const textObject: Record<string | symbol | number, any> = {}
  const lines = text.split('\n')
  lines.forEach((line: string) => {
    if (line !== '') {
      const [key, value] = line.split('=')
      if (key.trim() !== '' && value.trim() !== '')
        textObject[key.trim()] = value.trim()
    }
  })
  return textObject
}

export async function getFile({ path, enconding = 'utf-8' }: { path: string; enconding?: BufferEncoding }) {
  if ((path.includes('.js') || path.includes('.mjs')) && !path.includes('.json')) {
    const file = import(`${path}`)
    return file
  }
  return await readFile(path, enconding)
}

// helper function to find index of start of interface
export function findFileIndex(str: string, searchStr: string) {
  const index = str.indexOf(searchStr)
  return index !== -1 ? (index + searchStr.length - 1) : index
}

// helper function to find closing brace index
export function findClosingBraceIndex(str: string, start: number) {
  let count = 1
  for (let i = start + 1; i < str.length; i++) {
    if (str[i] === '{')
      count++
    if (str[i] === '}')
      count--
    if (count === 0)
      return i
  }
  return -1
}

/**
 * Check if searchIndex exist and inject code param inside the baces of the file
 * if searchIndex dont exist inject full code at the last line
 */
export function replaceOrCreateNewCode(fileData: string, searchIndex: string, code: string) {
  const start = findFileIndex(fileData, searchIndex)
  if (start === -1)
    return `${fileData}\n\n${searchIndex}\n${`${code}}`}`

  const openingBraceIndex = fileData.indexOf('{', start)
  if (openingBraceIndex === -1)
    throw new Error('Invalid interface definition')

  const end = findClosingBraceIndex(fileData, openingBraceIndex)
  return `${fileData.slice(0, openingBraceIndex + 1)}\n${`${code}}`}${fileData.slice(end + 1)}`
}

/**
 *
 * Read file and return or return empty string if file dont exist
 */
export async function readFileIgnoreError(filePath: string, enconding: 'utf-8' = 'utf-8'): Promise<string> {
  try {
    const data = await readFile(filePath, enconding)
    return data
  }
  catch (error: any) {
    if (error.code === 'ENOENT')
      return ''

    else throw error
  }
}
