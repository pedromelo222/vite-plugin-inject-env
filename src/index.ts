import { writeFile } from 'node:fs/promises'
import type { Plugin } from 'vite'
import { getFile, readFileIgnoreError, replaceOrCreateNewCode, transformStringToObject } from './file-utils'
import { importMeta } from './types'
import type { InjectPluginConfig, ParsedEnv, RecordEnv } from './types'

/**
 * Create vite env file to use as vite define config
 * { 'import.meta.env.VITE_APP_BASE_URL': '...' }
 */
export function parseEnv<T extends RecordEnv>(env: T): ParsedEnv<T> {
  const envTemp = {} as ParsedEnv<T>
  Object.keys(env).map(key => envTemp[`${importMeta}${key.toUpperCase()}` as keyof ParsedEnv<T>] = JSON.stringify(env[key]))
  return envTemp
}

/**
 * Returns array from single config or array config
 */
function createArrayFromConfig<T>(config: T[] | T | undefined) {
  return Array.isArray(config) ? [...config] : typeof config !== 'undefined' ? [config] : []
}

/**
 * Returns one array of modules and paths
 */
function getArrayFromConfig<T extends RecordEnv>({ module, path }: Pick<InjectPluginConfig<T>, 'module' | 'path'>) {
  return [...createArrayFromConfig(module), ...createArrayFromConfig(path)]
}

/**
 *  Return all envs from path and module into same object
 */
export async function loadEnvFile<T extends RecordEnv>({ module, path }: Pick<InjectPluginConfig<T>, 'module' | 'path'>) {
  const config = getArrayFromConfig({ module, path })
  const envObject: RecordEnv = Object.assign({}, ...await Promise.all(config.map(async (item) => {
    if (typeof item === 'string') {
      const file = await getFile({ path: item })
      if (item.includes('.md') || item.includes('.yml') || item.includes('.env'))
        return { ...parseEnv<T>(transformStringToObject(file)) }

      if (item.includes('.json'))
        return { ...parseEnv<T>(JSON.parse(file)) }
    }
    else if (typeof item === 'object') {
      return { ...parseEnv<T>(item) }
    }
  })))
  return envObject
}

/**
 * Create or edit env.d.ts to add type
 *
 */
async function replaceEnvTypeDefinition<T extends RecordEnv>(env: ParsedEnv<T> | undefined, filePath = './src/env.d.ts') {
  try {
    const data = await readFileIgnoreError(filePath)
    const newData = replaceOrCreateNewCode(data, 'interface ImportMetaEnv {', setInterfaceText(env))
    await writeFile(filePath, newData)
  }
  catch (err) {
    console.error(err)
  }
}

/**
 * Write type definition in string
 * "interface ImportMetaEnv { ... }"
 */
function setInterfaceText<T extends RecordEnv>(env: ParsedEnv<T> | undefined) {
  return Object.keys(env || {}).map(key => `  readonly ${key.split('import.meta.env.')[1]}: string\n`).join('')
}

/**
 *
 * vite-plugin-inject-env
 * inject enviroment variables into vite from '.md' '.yml' '.json' and 'js object'
 *
 */
export async function injectEnvPlugin<T extends RecordEnv>(pluginConfig: InjectPluginConfig<T>): Promise<Plugin> {
  const env = await loadEnvFile(pluginConfig)
  return {
    name: 'vite-inject-env-plugin',
    config: () => ({
      define: env,
    }),
    async configResolved() {
      if (pluginConfig.typeDefinition !== false) {
        try {
          await replaceEnvTypeDefinition(env)
        }
        catch (error: any) {
          throw new Error(`[TypeDefinition] - Error when creating types \n${error}`)
        }
      }
    },
  }
}
