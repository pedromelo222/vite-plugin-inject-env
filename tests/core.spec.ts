import { describe, expect, it } from 'vitest'
import { loadEnvFile } from '../src/index'

describe('core', () => {
  /**
     * .env file
     */
  it('should import envs from .env file', async () => {
    const env = await loadEnvFile({
      path: './tests/env-files/.env',
    })
    expect(env).toMatchSnapshot()
  })
  /**
     * md file
     */
  it('should import envs from MD file', async () => {
    const env = await loadEnvFile({
      path:
                './tests/env-files/md-file.md',
    })
    expect(env).toMatchSnapshot()
  })
  /**
     * yml file
     */
  it('should import envs from YML file', async () => {
    const env = await loadEnvFile({
      path: './tests/env-files/yml-file.yml',
    })
    expect(env).toMatchSnapshot()
  })
  /**
     * .json file
     */
  it('should import envs from JSON file', async () => {
    const env = await loadEnvFile({
      path: './tests/env-files/json-file.json',
    })
    expect(env).toMatchSnapshot()
  })
  /**
     *  javascript object
     */
  it('should import envs from a javascript object', async () => {
    const { envJs } = await import('./env-files/env')
    const env = await loadEnvFile({
      module: envJs,
    })
    expect(env).toMatchSnapshot()
  })
  /**
     *  array of files and modules
     */
  it('should import multiple envs files and modules', async () => {
    const { envJs, envJsTwo } = await import('./env-files/env')
    const env = await loadEnvFile({
      module: [
        envJs,
        envJsTwo,
      ],
      path: [
        './tests/env-files/yml-file.yml',
        './tests/env-files/md-file.md',
        './tests/env-files/json-file.json',
        './tests/env-files/.env',
      ],
    })
    expect(env).toMatchSnapshot()
  })
})
