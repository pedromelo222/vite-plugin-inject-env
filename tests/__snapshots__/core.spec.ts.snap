// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`core > should import envs from .env file 1`] = `
{
  "import.meta.env.VITE_APP_DOT_ENV_BASE_API_SOCKET_URL": "\\"wss://test.api.com\\"",
  "import.meta.env.VITE_APP_DOT_ENV_BASE_API_URL": "\\"https://testapi.com\\"",
  "import.meta.env.VITE_APP_DOT_ENV_BASE_AUTH_API": "\\"https://test.com\\"",
}
`;

exports[`core > should import envs from JSON file 1`] = `
{
  "import.meta.env.VITE_APP_JSON_ENV_BASE_API_SOCKET_URL": "\\"https://teste.com\\"",
  "import.meta.env.VITE_APP_JSON_ENV_BASE_API_URL": "\\"https://teste.com\\"",
  "import.meta.env.VITE_APP_JSON_ENV_BASE_AUTH_API": "\\"https://test.com\\"",
}
`;

exports[`core > should import envs from MD file 1`] = `
{
  "import.meta.env.VITE_APP_MD_ENV_BASE_API_SOCKET_URL": "\\"wss://test.api.com\\"",
  "import.meta.env.VITE_APP_MD_ENV_BASE_API_URL": "\\"https://testapi.com\\"",
  "import.meta.env.VITE_APP_MD_ENV_BASE_AUTH_API": "\\"https://test.com\\"",
}
`;

exports[`core > should import envs from YML file 1`] = `
{
  "import.meta.env.VITE_APP_YML_ENV_BASE_API_SOCKET_URL": "\\"wss://test.api.com\\"",
  "import.meta.env.VITE_APP_YML_ENV_BASE_API_URL": "\\"https://testapi.com\\"",
  "import.meta.env.VITE_APP_YML_ENV_BASE_AUTH_API": "\\"https://test.com\\"",
}
`;

exports[`core > should import envs from a javascript object 1`] = `
{
  "import.meta.env.VITE_APP_JS_BASE_API_SOCKET_URL": "\\"wss://test.api.com\\"",
  "import.meta.env.VITE_APP_JS_BASE_API_URL": "\\"https://testapi.com\\"",
  "import.meta.env.VITE_APP_JS_BASE_AUTH_API": "\\"https://test.com\\"",
}
`;

exports[`core > should import multiple envs files and modules 1`] = `
{
  "import.meta.env.VITE_APP_DOT_ENV_BASE_API_SOCKET_URL": "\\"wss://test.api.com\\"",
  "import.meta.env.VITE_APP_DOT_ENV_BASE_API_URL": "\\"https://testapi.com\\"",
  "import.meta.env.VITE_APP_DOT_ENV_BASE_AUTH_API": "\\"https://test.com\\"",
  "import.meta.env.VITE_APP_JSON_ENV_BASE_API_SOCKET_URL": "\\"https://teste.com\\"",
  "import.meta.env.VITE_APP_JSON_ENV_BASE_API_URL": "\\"https://teste.com\\"",
  "import.meta.env.VITE_APP_JSON_ENV_BASE_AUTH_API": "\\"https://test.com\\"",
  "import.meta.env.VITE_APP_JS_BASE_API_SOCKET_URL": "\\"wss://test.api.com\\"",
  "import.meta.env.VITE_APP_JS_BASE_API_URL": "\\"https://testapi.com\\"",
  "import.meta.env.VITE_APP_JS_BASE_AUTH_API": "\\"https://test.com\\"",
  "import.meta.env.VITE_APP_JS_TWO_BASE_API_SOCKET_URL": "\\"wss://test.api.com\\"",
  "import.meta.env.VITE_APP_JS_TWO_BASE_API_URL": "\\"https://testapi.com\\"",
  "import.meta.env.VITE_APP_JS_TWO_BASE_AUTH_API": "\\"https://test.com\\"",
  "import.meta.env.VITE_APP_MD_ENV_BASE_API_SOCKET_URL": "\\"wss://test.api.com\\"",
  "import.meta.env.VITE_APP_MD_ENV_BASE_API_URL": "\\"https://testapi.com\\"",
  "import.meta.env.VITE_APP_MD_ENV_BASE_AUTH_API": "\\"https://test.com\\"",
  "import.meta.env.VITE_APP_YML_ENV_BASE_API_SOCKET_URL": "\\"wss://test.api.com\\"",
  "import.meta.env.VITE_APP_YML_ENV_BASE_API_URL": "\\"https://testapi.com\\"",
  "import.meta.env.VITE_APP_YML_ENV_BASE_AUTH_API": "\\"https://test.com\\"",
}
`;
